sw/device/lib/msg_uart/msg_impl.h

```diff
@@ -0,0 +1,56 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef _MSG_IMPL_H
+#define _MSG_IMPL_H
+
+#include "sw/device/lib/msg_print.h"
+#include "sw/device/lib/uart.h"
+
+// Stringify stuff.
+#define _xstr(a) _str(a)
+#define _str(a) #a
+
+// The macro below allows forwarding arguments from a variadic macro to a
+// variadic function and support the case where number of arguments is 0.
+#define MSG_VA_ARGS(...) , ##__VA_ARGS__
+
+/**
+ * Message type and verbosity string constants
+ */
+#define MSG_TYPE_INFO "INFO"
+#define MSG_TYPE_WARNING "WARNING"
+#define MSG_TYPE_ERROR "ERROR"
+#define MSG_TYPE_FATAL "FATAL"
+
+/**
+ * Message header helper macro
+ *
+ * This macro constructs the message header in the following way:
+ * If MSG_HEADER_INCL_FILE_LINE flag is set:      TYPE (file:line):
+ * If MSG_HEADER_INCL_FILE_LINE flag is not set:  TYPE:
+ *
+ * Message verbosity is currently unused, since the volume of messages
+ * is not expected to be very high.
+ */
+#ifdef MSG_HEADER_INCL_FILE_LINE
+#define MSG_HEADER(msg_type, verbosity) \
+  msg_type verbosity " (" __FILE__ ":" _xstr(__LINE__) "): "
```

Comment:
To improve error messages, you can write
```c
"" msg_type "" verbosity " ... "
```
This helps the compiler make sure that msg_type and verbosity are string literals.

Reply:
Done.